stat/mds/mds.go

```diff
@@ -0,0 +1,83 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package mds
+
+import (
+	"math"
+
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/mat"
+)
+
+// TorgersonScaling converts a dissimilarity matrix to a matrix containing
+// Euclidean coordinates. TorgersonScaling places the coordinates in dst and
+// returns it and true if successful. If the scaling is not successful, dst
+// is returned, but will not be a valid scaling.
+//
+// If dst is nil, a new mat.Dense is allocated. If dst is not a zero matrix,
+// the dimensions of dst and dis must match otherwise TorgersonScaling will panic.
+// The dis matrix must be square or TorgersonScaling will panic.
+func TorgersonScaling(dst *mat.Dense, dis mat.Symmetric) (mds *mat.Dense, ok bool) {
+	// https://doi.org/10.1007/0-387-28981-X_12
+
+	n := dis.Symmetric()
+	if dst == nil {
+		dst = mat.NewDense(n, n, nil)
+	} else if r, c := dst.Dims(); !dst.IsZero() && (r != n || c != n) {
+		panic(mat.ErrShape)
+	}
+
+	b := mat.NewSymDense(n, nil)
+	for i := 0; i < n; i++ {
+		for j := i; j < n; j++ {
+			v := dis.At(i, j)
+			v *= v
+			b.SetSym(i, j, v)
+		}
+	}
+	c := mat.NewSymDense(n, nil)
+	s := -1 / float64(n)
+	for i := 0; i < n; i++ {
+		c.SetSym(i, i, 1+s)
+		for j := i + 1; j < n; j++ {
+			c.SetSym(i, j, s)
+		}
+	}
+	dst.Product(c, b, c)
+	for i := 0; i < n; i++ {
+		for j := i; j < n; j++ {
+			b.SetSym(i, j, -0.5*dst.At(i, j))
+		}
+	}
+
+	var ed mat.EigenSym
+	ok = ed.Factorize(b, true)
+	if !ok {
+		return dst, false
+	}
+	dst.EigenvectorsSym(&ed)
+	vals := ed.Values(nil)
+	reverse(vals, dst.RawMatrix())
+	for i, v := range vals {
+		if v < 0 {
+			vals[i] = 0
+			continue
+		}
+		vals[i] = math.Sqrt(v)
+	}
+	dst.Mul(dst, mat.NewDiagonal(len(vals), vals))
```

Comment:
If m is the number of negative eigenvalues, then `dst` will have last m columns zero. Isn't this information worth returning somehow?

Reply:
We could return the Eigenvalues. Some packages in other languages off this as an option.

Reply:
All of them or only the positive?

Reply:
Don't know, probably all of them, but we could slice to the positive only (this would make the negative values available by slicing to `cap(v)` (which we'd have to document).

Reply:
Not sure how much value the negative eigenvalues have, so slicing to the positive values only seems better, also so that the number of non-zero columns is more readily available. Slicing to `cap(v)` is an interesting idea but isn't it a bit unconventional?

Reply:
>Slicing to `cap(v)` is an interesting idea but isn't it a bit unconventional?

I think it's horrific.

I think adding an input destination slice and returning it is probably the way to go (allocating if it's `nil`). We check that it's adequately long (or that it's exactly correct length for the input - any preference?) and slice to include only the positive elements and return it.

Reply:
If the eigenvalues are not extremely useful, then just returning the number of non-zero columns would be the easiest. If they are useful, then adding a destination slice sound good to me. Or, having a dedicated type for this?

Reply:
The API here is tricky. I think the simplest thing to do is take a Eigenvalues destination slice and return it and the number of positive values (probably in place of the `ok bool`). Then the user can slice to the correct length if they want (that's essentially what we're getting them to do with the `mds` value too).

Reply:
Just to be clear, by "returning the number of non-zero columns" I mean having a `cols int` return.

The simplest thing would be not take the destinations and just allocate and return, but since we want to reuse, then the destination slice is probably the simplest.

Reply:
Yeah, what I've done is perhaps odd, but please take a look anyway.